src/irr/asset/CCPUMeshPacker.h

```diff
@@ -46,85 +67,243 @@ PackedMeshBufferData CCPUMeshPacker<MDIStructType>::alloc(const ICPUMeshBuffer c
+		const uint32_t minIdxVal = *std::min_element(idxBuffBatchBegin, idxBuffBatchEnd);
+
+		MDIStructType MDIData;
+		MDIData.count = idxBuffBatchEnd - idxBuffBatchBegin;
+		MDIData.instanceCount = inputMeshBuffer->getInstanceCount();
+		MDIData.firstIndex = firstIdxForCurrMeshBatch;
+		MDIData.baseVertex = *std::min_element(idxBuffBatchBegin, idxBuffBatchEnd);
+		MDIData.baseInstance = 0u; /*TODO*/
+
+		//set mdi structs
+		memcpy(mdiBuffPtr, &MDIData, sizeof(MDIStructType));
+		
+		//set indices
+		//TODO?: if constexpr for EIT
+		for (IndexType* ptr = idxBuffBatchBegin; ptr != idxBuffBatchEnd; ptr++)
 		{
-			idxCnt = maxIdxCntPerPatch;
+			*indexBuffPtr = *ptr - minIdxVal;
+			indexBuffPtr++;
 		}
 
-		*(mdiBuffPtr + MDIAllocAddr + i) =
-		{
-			idxCnt,
-			static_cast<uint32_t>(meshBuffer->getInstanceCount()),
-			idxAllocAddr + i * maxIdxCntPerPatch,
-			0 /*TODO*/,
-			static_cast<uint32_t>(meshBuffer->getBaseInstance()) /*TODO*/
-		};
-	}
+		firstIdxForCurrMeshBatch += MDIData.count;
+		idxBuffBatchBegin = idxBuffBatchEnd;
 
+		mdiBuffPtr++;
+		indexBuffPtr += MDIData.count;
+	}
 
-	PackedMeshBufferData result{ MDIAllocAddr, MDIStructsNeeded };
-	return result;
+	return {};
 }
 
 template <typename MDIStructType>
-void CCPUMeshPacker<MDIStructType>::commit()
+template <typename Iterator>
+PackedMeshBufferData CCPUMeshPacker<MDIStructType>::commit(const Iterator begin, const Iterator end, ReservedAllocationMeshBuffers ramb)
 {
-	return;
+	//AFTER SAFE SHRINK FIX TODO LIST:
+	//multiple ReservedAllocationMeshbuffers on input
+	//new size for buffers (obviously)
+	//fix shit, so Iterator may be an iterator to core::smart_refct_ptr too
+	//handle case where (maxIdx - minIdx) > 0xFFFF 
+
+	uint8_t* perVertexBuffPtr = static_cast<uint8_t*>(outputBuffer.perVertexVtxBuffer.buffer->getPointer()) + (ramb.vertexAllocationOffset * m_vtxSize);
+	uint8_t* perInstVtxBuffPtr = static_cast<uint8_t*>(outputBuffer.perInstanceVtxBuffer.buffer->getPointer()) + (ramb.instanceAllocationOffset * m_perInstVtxSize);
+
+		//deinterleave vertices
+	//I think I've should validate input mesh buffers again..
+	for (uint16_t attrBit = 0x0001, location = 0; location < SVertexInputParams::MAX_ATTR_BUF_BINDING_COUNT; attrBit <<= 1, location++)
+	{
+		if (!(attrBit & m_outVtxInputParams.enabledAttribFlags))
+			continue;
+
+		//TODO: if, for case where currently processed mesh buffer doesn't have given attr (important!)
+		
+		SVertexInputAttribParams attrib = m_outVtxInputParams.attributes[location];
+		SVertexInputBindingParams attribBinding = m_outVtxInputParams.bindings[attrib.binding];
+
+		const size_t attrSize = asset::getTexelOrBlockBytesize(static_cast<E_FORMAT>(attrib.format));
+		const size_t stride = (attribBinding.stride) == 0 ? attrSize : attribBinding.stride;
+
+			//this is where vertices are deinterleaved and copied into output vertex buffers
+		for (auto it = begin; it != end; it++)
```

Comment:
hoist this loop above the vertex attribute loop, respect data locality and your CPUs caches.

Iterate/Loop from most sparse data to most packed.. So

MeshBuffer > Indices > Vertex Attributes